### PR TITLE
[rs] 07: Moving all rendering, windowing and window input related code into the renderer crate

### PR DESCRIPTION
**Pull request by Rhuagh**

### CODE REVIEW

File: amethyst_renderer/src/pass/flat/interleaved.rs [hunk 1]
```diff
@@ -23,16 +24,14 @@ use super::*;
 
 /// Draw mesh without lighting
 /// `V` is `VertexFormat`
-/// `T` is transform matrix component
 #[derive(Clone, Debug, PartialEq)]
-pub struct DrawFlat<V, T> {
-    _pd: PhantomData<(V, T)>,
+pub struct DrawFlat<V> {
+    _pd: PhantomData<V>,
```
Comment (jojolepro) [inline amethyst_renderer/src/pass/flat/interleaved.rs]: :+1:

File: amethyst_renderer/Cargo.toml [hunk 2]
```diff
@@ -66,3 +69,7 @@ version = "0.10"
 
 [dev-dependencies]
 genmesh = "0.4"
+
+[dependencies.smallvec]
+version = "0.4.2"
+features = ["serde"]
```
Comment (jojolepro) [inline amethyst_renderer/Cargo.toml]: Missing \n

File: amethyst_renderer/Cargo.toml [hunk 3]
```diff
@@ -39,6 +40,8 @@ rayon = "0.8"
 rayon-core = "1.2"
 serde = "1.0"
 serde_derive = "1.0"
+shred = "0.5"
+shrev = "0.6"
```
Comment (jojolepro) [inline amethyst_renderer/Cargo.toml]: Intended duping?

File: amethyst_renderer/src/system.rs [hunk 4]
```diff
@@ -3,32 +3,56 @@
...
+use error::Result;
 
 /// Rendering system.
 #[derive(Derivative)]
 #[derivative(Debug)]
 pub struct RenderSystem<P> {
     pipe: P,
-    #[derivative(Debug = "ignore")]
-    renderer: Renderer,
+    #[derivative(Debug = "ignore")] renderer: Renderer,
```
Comment (jojolepro) [inline amethyst_renderer/src/system.rs]: format

File: examples/03_renderable/main.rs [hunk 5]
```diff
@@ -363,17 +363,14 @@ fn run() -> Result<(), Error> {
         .with::<ExampleSystem>(ExampleSystem, "example_system", &[])
         .with_bundle(TransformBundle::new().with_dep(&["example_system"]))?
         .with_bundle(RenderBundle::new())?
-        .with_local(create_render_system(
-            pipeline_builder,
-            Some(display_config),
-        )?)
+        .with_local(RenderSystem::build(pipeline_builder, Some(display_config))?)
```
Comment (jojolepro) [inline examples/03_renderable/main.rs]: Doesn't really matter, but here "pipeline_builder" is used, while in the pong example it is "pipe".

File: src/audio/output.rs [hunk 6]
```diff
@@ -26,8 +26,7 @@ impl Output {
     /// Play a sound once.  A volume of 1.0 is unchanged, while 0.0 is silent.
     ///
     /// This will return an Error if the loaded audio file in source could not be decoded.
-    pub fn try_play_once(&self, source: &Source, volume: f32)
-        -> Result<(), DecoderError> {
+    pub fn try_play_once(&self, source: &Source, volume: f32) -> Result<(), DecoderError> {
```
Comment (jojolepro) [inline src/audio/output.rs]: Its weird how sometimes it wants it on one line, and sometimes not.

File: src/ecs/audio/systems/audio.rs [hunk 7]
```diff
@@ -5,11 +5,12 @@ use std::sync::atomic::{AtomicBool, Ordering};
 use std::time::Duration;
 
 use cgmath::{Matrix4, Point3, Transform};
+use core::transform::Transform as TransformComponent;
 use rodio::{Sample, Source, SpatialSink};
 
 use ecs::{Entity, Fetch, Join, ReadStorage, System, WriteStorage};
 use ecs::audio::components::{AudioEmitter, AudioListener};
-use ecs::transform::Transform as TransformComponent;
+
```
Comment (jojolepro) [inline src/ecs/audio/systems/audio.rs]: Formatter dying internally at this point.

File: amethyst_renderer/Cargo.toml (hunk 3 above)
Comment (Xaeroxe) [reply amethyst_renderer/Cargo.toml]: Not a dupe!  Those are different crates.

Comment (jojolepro) [reply amethyst_renderer/Cargo.toml]: I'll stop reviewing, I'm starting to see different things as the same now. :(

Comment (Xaeroxe) [reply amethyst_renderer/Cargo.toml]: Nah, don't worry, to `Err` is human :P

**Comment by Rhuagh** [comment]
As the window is private to the renderer it doesn't make sense to put it anywhere else

**Comment by omni-viral** [comment]
Well. Maybe it shouldn't

**Comment by Rhuagh** [comment]
I happen to agree with you on that, but I think that's outside the scope of this PR.

File: amethyst_renderer/src/system.rs (hunk 4 above)
Comment (Rhuagh) [reply amethyst_renderer/src/system.rs]: stupid rustfmt :(

File: src/audio/output.rs (hunk 6 above)
Comment (Rhuagh) [reply src/audio/output.rs]: Yup.

File: src/ecs/audio/systems/audio.rs (hunk 7 above)
Comment (Rhuagh) [reply src/ecs/audio/systems/audio.rs]: What's the problem here? The Transform component has moved to core, so it's now `core::transform::Transform` a few lines above ?

File: examples/03_renderable/main.rs (hunk 5 above)
Comment (Rhuagh) [reply examples/03_renderable/main.rs]: Fixing naming inconsistencies feels like something outside the scope of this PR.

**Comment by Rhuagh** [comment]
Review comments addressed, please review again @jojolepro.

File: src/ecs/audio/systems/audio.rs (hunk 7 above)
Comment (jojolepro) [reply src/ecs/audio/systems/audio.rs]: It decided to have two spaces inside of one. It doesn't show well on the preview though.

**Comment by jojolepro** [comment]
Will do soon :tm:

**Comment by Xaeroxe** [comment]
Handled in #430

**Pull request closed without merging**